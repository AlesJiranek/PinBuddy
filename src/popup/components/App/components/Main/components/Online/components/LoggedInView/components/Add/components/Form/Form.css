.form {

}

.form__options {
  display: flex;
}

.form__option {
  margin-right: var(--lh);
}
