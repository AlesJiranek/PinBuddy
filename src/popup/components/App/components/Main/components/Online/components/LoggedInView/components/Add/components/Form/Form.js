import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { postsAdd } from 'redux-popup/actions/posts';
import { Input, Button, Checkbox } from 'theme';
import './Form.css';

class Form extends Component {
  state ={
    title: '',
    url: '',
    description: '',
    tags: '',
    privatePost: false,
    readLater: false,
  };

  refInput = React.createRef();

  componentDidMount() {

    chrome.tabs.query({ active: true }, result => {
      const { title, url } = result[0];
      this.setState({
        title,
        url,
      });
    });

    chrome.tabs.executeScript({ code: 'window.getSelection().toString()' }, selection => {
      if (selection) {
        this.setState({
          description: selection[0],
        });
      }
    });

    chrome.storage.sync.get(['privateCheckboxByDefault', 'toReadChecboxByDefault'], result => {
      this.setState({
        privatePost: result.privateCheckboxByDefault,
        readLater: result.toReadChecboxByDefault,
      });
    });

    window.addEventListener('keydown', this.handleKeydown);

    this.refInput.current.focus();
  }

  componentWillUnmount() {
    window.removeEventListener('keydown', this.handleKeydown);
  }

  render() {
    const formInvalid =
      this.state.title.length > 255 ||
      this.state.description.length > 65536 ||
      this.state.tags.length > 255;

    return (
      <div className="form">

        <Input
          id="title"
          label={chrome.i18n.getMessage('popupAddTitleLabel')}
          placeholder={chrome.i18n.getMessage('popupAddTitleLabel')}
          value={this.state.title}
          onChange={this.handleInputChange}
          pattern=".{1,255}"
          invalidMessage={chrome.i18n.getMessage('popupAddTitleInvalidMessage')}
        />

        <Input
          id="description"
          label={chrome.i18n.getMessage('popupAddDescriptionLabel')}
          placeholder={chrome.i18n.getMessage('popupAddDescriptionLabel')}
          value={this.state.description}
          onChange={this.handleInputChange}
          pattern=".{1,65536}"
          invalidMessage={chrome.i18n.getMessage('popupAddDescriptionInvalidMessage')}
        />

        <Input
          id="tags"
          label={chrome.i18n.getMessage('popupAddTagsLabel')}
          placeholder={chrome.i18n.getMessage('popupAddTagsLabel')}
          value={this.state.tags}
          onChange={this.handleInputChange}
          ref={this.refInput}
          pattern=".{1,255}"
          invalidMessage={chrome.i18n.getMessage('popupAddTagsInvalidMessage')}
        />

        <div className="form__options">
          <div className="form__option">

            <Checkbox
              id="privatePost"
              label={chrome.i18n.getMessage('popupAddPrivateCheckboxLabel')}
              checked={this.state.privatePost}
              onChange={this.handleCheckboxChange}
            />

          </div>
          <div className="form__option">

            <Checkbox
              id="readLater"
              label={chrome.i18n.getMessage('popupAddReadLaterCheckboxLabel')}
              checked={this.state.readLater}
              onChange={this.handleCheckboxChange}
            />

          </div>
        </div>

        <Button
          t={chrome.i18n.getMessage('popupAddBookmarkButton')}
          title={`${chrome.i18n.getMessage('popupAddBookmarkButton')} ${navigator.platform === 'MacIntel' ? '(⌘ + Enter)' : '(Ctrl + Enter)'}`}
          disabled={formInvalid}
          onClick={this.handleButtonClick}
        />

      </div>
    );
  }

  handleInputChange = e => {
    const { id, value } = e.target;
    this.setState({
      [id]: value,
    });
  }

  handleCheckboxChange = e => {
    const { id, checked } = e.target;
    this.setState({
      [id]: checked,
    });
  }

  handleButtonClick = () => {
    this.setState({
      laoding: true,
    });
    this.props.postsAdd(this.state);
  }

  handleKeydown = e => {
    // ⌘ + Enter
    // TODO: Check if this is working on Windows
    if (e.keyCode === 13 && (e.ctrlKey || e.metaKey)) {
      this.props.postsAdd(this.state);
    }
  }
}

Form.propTypes = {
  postsAdd: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  postsAdd: postInfo => dispatch(postsAdd(postInfo)),
});

export default connect(null, mapDispatchToProps)(Form);
