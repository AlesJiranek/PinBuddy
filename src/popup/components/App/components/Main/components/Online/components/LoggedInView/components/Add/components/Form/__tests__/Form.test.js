import React from 'react';
import { shallow } from 'enzyme';

import { Form } from '../Form';

describe('<Form />', () => {
  const defaultProps = {
    postsAdd: jest.fn(),
  };

  it('should render correctly', () => {
    const tree = shallow(
      <Form
        {...defaultProps}
      />
    );
    expect(tree).toMatchSnapshot();
  });

});
