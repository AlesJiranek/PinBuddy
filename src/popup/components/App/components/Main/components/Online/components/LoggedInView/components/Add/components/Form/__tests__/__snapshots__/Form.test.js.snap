// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    postsAdd={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ForwardRef
          autoCapitalize={null}
          autoComplete={null}
          autoCorrect={null}
          autoFocus={null}
          id="title"
          invalidMessage="popupAddTitleInvalidMessage"
          label="popupAddTitleLabel"
          onChange={[Function]}
          pattern=".{1,255}"
          placeholder="popupAddTitleLabel"
          spellCheck={null}
          type="text"
          value=""
        />,
        <ForwardRef
          autoCapitalize={null}
          autoComplete={null}
          autoCorrect={null}
          autoFocus={null}
          id="description"
          invalidMessage="popupAddDescriptionInvalidMessage"
          label="popupAddDescriptionLabel"
          onChange={[Function]}
          pattern=".{1,65536}"
          placeholder="popupAddDescriptionLabel"
          spellCheck={null}
          type="text"
          value=""
        />,
        <ForwardRef
          autoCapitalize={null}
          autoComplete={null}
          autoCorrect={null}
          autoFocus={true}
          id="tags"
          invalidMessage="popupAddTagsInvalidMessage"
          label="popupAddTagsLabel"
          onChange={[Function]}
          pattern=".{1,255}"
          placeholder="popupAddTagsLabel"
          spellCheck={null}
          type="text"
          value=""
        />,
        <div
          className="form__options"
        >
          <div
            className="form__option"
          >
            <Checkbox
              checked={false}
              id="privatePost"
              label="popupAddPrivateCheckboxLabel"
              onChange={[Function]}
            />
          </div>
          <div
            className="form__option"
          >
            <Checkbox
              checked={false}
              id="readLater"
              label="popupAddReadLaterCheckboxLabel"
              onChange={[Function]}
            />
          </div>
        </div>,
        <Button
          disabled={false}
          onClick={[Function]}
          secondary={false}
          t="popupAddBookmarkButton"
          title="popupAddBookmarkButton (Ctrl + Enter)"
          type="button"
        />,
      ],
      "className": "form",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "autoCapitalize": null,
          "autoComplete": null,
          "autoCorrect": null,
          "autoFocus": null,
          "id": "title",
          "invalidMessage": "popupAddTitleInvalidMessage",
          "label": "popupAddTitleLabel",
          "onChange": [Function],
          "pattern": ".{1,255}",
          "placeholder": "popupAddTitleLabel",
          "spellCheck": null,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "autoCapitalize": null,
            "autoComplete": null,
            "autoCorrect": null,
            "autoFocus": null,
            "invalidMessage": null,
            "pattern": null,
            "spellCheck": null,
            "type": "text",
          },
          "propTypes": Object {
            "autoCapitalize": [Function],
            "autoComplete": [Function],
            "autoCorrect": [Function],
            "autoFocus": [Function],
            "id": [Function],
            "invalidMessage": [Function],
            "label": [Function],
            "onChange": [Function],
            "pattern": [Function],
            "spellCheck": [Function],
            "type": [Function],
          },
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "autoCapitalize": null,
          "autoComplete": null,
          "autoCorrect": null,
          "autoFocus": null,
          "id": "description",
          "invalidMessage": "popupAddDescriptionInvalidMessage",
          "label": "popupAddDescriptionLabel",
          "onChange": [Function],
          "pattern": ".{1,65536}",
          "placeholder": "popupAddDescriptionLabel",
          "spellCheck": null,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "autoCapitalize": null,
            "autoComplete": null,
            "autoCorrect": null,
            "autoFocus": null,
            "invalidMessage": null,
            "pattern": null,
            "spellCheck": null,
            "type": "text",
          },
          "propTypes": Object {
            "autoCapitalize": [Function],
            "autoComplete": [Function],
            "autoCorrect": [Function],
            "autoFocus": [Function],
            "id": [Function],
            "invalidMessage": [Function],
            "label": [Function],
            "onChange": [Function],
            "pattern": [Function],
            "spellCheck": [Function],
            "type": [Function],
          },
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "autoCapitalize": null,
          "autoComplete": null,
          "autoCorrect": null,
          "autoFocus": true,
          "id": "tags",
          "invalidMessage": "popupAddTagsInvalidMessage",
          "label": "popupAddTagsLabel",
          "onChange": [Function],
          "pattern": ".{1,255}",
          "placeholder": "popupAddTagsLabel",
          "spellCheck": null,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "autoCapitalize": null,
            "autoComplete": null,
            "autoCorrect": null,
            "autoFocus": null,
            "invalidMessage": null,
            "pattern": null,
            "spellCheck": null,
            "type": "text",
          },
          "propTypes": Object {
            "autoCapitalize": [Function],
            "autoComplete": [Function],
            "autoCorrect": [Function],
            "autoFocus": [Function],
            "id": [Function],
            "invalidMessage": [Function],
            "label": [Function],
            "onChange": [Function],
            "pattern": [Function],
            "spellCheck": [Function],
            "type": [Function],
          },
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="form__option"
            >
              <Checkbox
                checked={false}
                id="privatePost"
                label="popupAddPrivateCheckboxLabel"
                onChange={[Function]}
              />
            </div>,
            <div
              className="form__option"
            >
              <Checkbox
                checked={false}
                id="readLater"
                label="popupAddReadLaterCheckboxLabel"
                onChange={[Function]}
              />
            </div>,
          ],
          "className": "form__options",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Checkbox
                checked={false}
                id="privatePost"
                label="popupAddPrivateCheckboxLabel"
                onChange={[Function]}
              />,
              "className": "form__option",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "checked": false,
                "id": "privatePost",
                "label": "popupAddPrivateCheckboxLabel",
                "onChange": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Checkbox
                checked={false}
                id="readLater"
                label="popupAddReadLaterCheckboxLabel"
                onChange={[Function]}
              />,
              "className": "form__option",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "checked": false,
                "id": "readLater",
                "label": "popupAddReadLaterCheckboxLabel",
                "onChange": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "disabled": false,
          "onClick": [Function],
          "secondary": false,
          "t": "popupAddBookmarkButton",
          "title": "popupAddBookmarkButton (Ctrl + Enter)",
          "type": "button",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ForwardRef
            autoCapitalize={null}
            autoComplete={null}
            autoCorrect={null}
            autoFocus={null}
            id="title"
            invalidMessage="popupAddTitleInvalidMessage"
            label="popupAddTitleLabel"
            onChange={[Function]}
            pattern=".{1,255}"
            placeholder="popupAddTitleLabel"
            spellCheck={null}
            type="text"
            value=""
          />,
          <ForwardRef
            autoCapitalize={null}
            autoComplete={null}
            autoCorrect={null}
            autoFocus={null}
            id="description"
            invalidMessage="popupAddDescriptionInvalidMessage"
            label="popupAddDescriptionLabel"
            onChange={[Function]}
            pattern=".{1,65536}"
            placeholder="popupAddDescriptionLabel"
            spellCheck={null}
            type="text"
            value=""
          />,
          <ForwardRef
            autoCapitalize={null}
            autoComplete={null}
            autoCorrect={null}
            autoFocus={true}
            id="tags"
            invalidMessage="popupAddTagsInvalidMessage"
            label="popupAddTagsLabel"
            onChange={[Function]}
            pattern=".{1,255}"
            placeholder="popupAddTagsLabel"
            spellCheck={null}
            type="text"
            value=""
          />,
          <div
            className="form__options"
          >
            <div
              className="form__option"
            >
              <Checkbox
                checked={false}
                id="privatePost"
                label="popupAddPrivateCheckboxLabel"
                onChange={[Function]}
              />
            </div>
            <div
              className="form__option"
            >
              <Checkbox
                checked={false}
                id="readLater"
                label="popupAddReadLaterCheckboxLabel"
                onChange={[Function]}
              />
            </div>
          </div>,
          <Button
            disabled={false}
            onClick={[Function]}
            secondary={false}
            t="popupAddBookmarkButton"
            title="popupAddBookmarkButton (Ctrl + Enter)"
            type="button"
          />,
        ],
        "className": "form",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "autoCapitalize": null,
            "autoComplete": null,
            "autoCorrect": null,
            "autoFocus": null,
            "id": "title",
            "invalidMessage": "popupAddTitleInvalidMessage",
            "label": "popupAddTitleLabel",
            "onChange": [Function],
            "pattern": ".{1,255}",
            "placeholder": "popupAddTitleLabel",
            "spellCheck": null,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "autoCapitalize": null,
              "autoComplete": null,
              "autoCorrect": null,
              "autoFocus": null,
              "invalidMessage": null,
              "pattern": null,
              "spellCheck": null,
              "type": "text",
            },
            "propTypes": Object {
              "autoCapitalize": [Function],
              "autoComplete": [Function],
              "autoCorrect": [Function],
              "autoFocus": [Function],
              "id": [Function],
              "invalidMessage": [Function],
              "label": [Function],
              "onChange": [Function],
              "pattern": [Function],
              "spellCheck": [Function],
              "type": [Function],
            },
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "autoCapitalize": null,
            "autoComplete": null,
            "autoCorrect": null,
            "autoFocus": null,
            "id": "description",
            "invalidMessage": "popupAddDescriptionInvalidMessage",
            "label": "popupAddDescriptionLabel",
            "onChange": [Function],
            "pattern": ".{1,65536}",
            "placeholder": "popupAddDescriptionLabel",
            "spellCheck": null,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "autoCapitalize": null,
              "autoComplete": null,
              "autoCorrect": null,
              "autoFocus": null,
              "invalidMessage": null,
              "pattern": null,
              "spellCheck": null,
              "type": "text",
            },
            "propTypes": Object {
              "autoCapitalize": [Function],
              "autoComplete": [Function],
              "autoCorrect": [Function],
              "autoFocus": [Function],
              "id": [Function],
              "invalidMessage": [Function],
              "label": [Function],
              "onChange": [Function],
              "pattern": [Function],
              "spellCheck": [Function],
              "type": [Function],
            },
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "autoCapitalize": null,
            "autoComplete": null,
            "autoCorrect": null,
            "autoFocus": true,
            "id": "tags",
            "invalidMessage": "popupAddTagsInvalidMessage",
            "label": "popupAddTagsLabel",
            "onChange": [Function],
            "pattern": ".{1,255}",
            "placeholder": "popupAddTagsLabel",
            "spellCheck": null,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "autoCapitalize": null,
              "autoComplete": null,
              "autoCorrect": null,
              "autoFocus": null,
              "invalidMessage": null,
              "pattern": null,
              "spellCheck": null,
              "type": "text",
            },
            "propTypes": Object {
              "autoCapitalize": [Function],
              "autoComplete": [Function],
              "autoCorrect": [Function],
              "autoFocus": [Function],
              "id": [Function],
              "invalidMessage": [Function],
              "label": [Function],
              "onChange": [Function],
              "pattern": [Function],
              "spellCheck": [Function],
              "type": [Function],
            },
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="form__option"
              >
                <Checkbox
                  checked={false}
                  id="privatePost"
                  label="popupAddPrivateCheckboxLabel"
                  onChange={[Function]}
                />
              </div>,
              <div
                className="form__option"
              >
                <Checkbox
                  checked={false}
                  id="readLater"
                  label="popupAddReadLaterCheckboxLabel"
                  onChange={[Function]}
                />
              </div>,
            ],
            "className": "form__options",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Checkbox
                  checked={false}
                  id="privatePost"
                  label="popupAddPrivateCheckboxLabel"
                  onChange={[Function]}
                />,
                "className": "form__option",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "checked": false,
                  "id": "privatePost",
                  "label": "popupAddPrivateCheckboxLabel",
                  "onChange": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Checkbox
                  checked={false}
                  id="readLater"
                  label="popupAddReadLaterCheckboxLabel"
                  onChange={[Function]}
                />,
                "className": "form__option",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "checked": false,
                  "id": "readLater",
                  "label": "popupAddReadLaterCheckboxLabel",
                  "onChange": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "disabled": false,
            "onClick": [Function],
            "secondary": false,
            "t": "popupAddBookmarkButton",
            "title": "popupAddBookmarkButton (Ctrl + Enter)",
            "type": "button",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
